Finished tutorials from UW-Madison Tutorial Code Sprinting Workshop

eblur opened the pull request:
Finished tutorials from the UW-Madison Astropy Tutorial Code Sprinting Workshop (Sept 28-29).

This PR replaces #162, #163, #165, #166. From those PR comments, I pulled in suggestions for minor changes that don't require writing additional content.

Tutorials that needed more extensive changes will be addressed in subsequent PRs.

eblur commented:
Need to change Coordinates-Transform.ipynb to catch error so it doesn't break Travis

eblur commented:
There has been discussion around the [new **astroquery** release (0.3.7) that was required to fix a bug in ESASky](https://github.com/astropy/astroquery/issues/1021), used in the FITS-cubes tutorial. The release is not yet available with conda package management.

It has been suggested that we change the yaml files so that Travis uses pip instead. Common practice is to instruct new users to implement conda for all of their package management needs. Changing to pip would require a notice at the top of the tutorial informing users of the patched astroquery version and installation instructions. Newer users are likely to get tripped up by this. **Please comment on this potential change.**

This may not be urgent. Is there an ETA for the conda update to astroquery 0.3.7?

@keflavich @eteq @kelle @bsipocz @pllim @vditomasso @astrofrog

keflavich commented:
Where is the FITS-cubes tutorial?  I'm confused why users would be exposed to differences between conda and pip if there are changes in the travis.yml file.

I don't really understand how/when conda updates from pypi; someone else should answer that.

bsipocz commented:
the conda package is being built atm, depending on the travis queue, it may take up to a few days to get available on the astropy channel.

mwcraig commented:
First, thanks to @bsipocz for updating the conda package; like she said it should be ready soon. **Update:** it is available now!

> Best practice is using conda for package management over pip, whenever possible, to take advantage of conda environments. Do people agree?

I may be the wrong person to answer because I'm strongly biased towards conda, but these are what I see as the advantages of conda for package management:

- Easier to provide binary builds for all major platforms.
- uniform instructions for creating/managing environment across all major platforms (wasn't entirely the case until a relatively recent release of conda but is true now).
- Good support for build tools from anaconda (the company) and the conda-forge community.
- No compilers need to be installed by/for users/
- Install in a user's home directory instead of system directory (unless you force it to install system wide and have permissions to do so).

The cross-platform nature of conda is particularly important to me because many of the undergrads I work with have a Windows machine and compilation is...painful, at best.

> Is pip truly interchangeable with conda? Will conda environments, etc, be preserved if some one has to install a package using pip? @mwcraig ?

Conda environments are fine with packages installed with either pip or conda, with a few caveats:

- For a given package you should pick pip or conda and stick with it. conda gets confused if you both conda install and pip install the same package, especially if they are different versions.
- Mixing the default conda channel (from anaconda the company) and conda-forge can, in fairly rare cases, cause conflicts. I have not received any reports for astropy-related packages. 
- Some times you have to install with pip because there is no conda package. 

So I wouldn't say they are interchangable, and I wouldn't even recommend that users immediately try pip if conda doesn't work. But if the package is really only available via pip, then that is the way to install it, and the conda environment will be aware that it is installed.

mwcraig commented:
Spoke a little too soon...the linux and mac builds of 0.3.7 are done, but Windows will take a little longer (done by morning though).

bsipocz commented:
@mwcraig - you're overly optimistic, the appveyor queue is very jammed (the current jobs that are running were submitted a day ago as part of a large batch of PRs to staged-recipes), and even when merged it will take the same amount of time again to get the build, and then wait for the next day cron to copy it over. So I would say don't expect much before the end of the week...

mwcraig commented:
:crying:

eblur commented:
Thanks @mwcraig for taking the time to answer these questions in detail. 

For the purposes of this astroquery example, we will refrain from telling users to install with pip, since it is also available in conda.

mwcraig commented:
fyi, we are working on resolving a technical issue with the astroquery 0.3.7 builds, hopefully resolved soon (I think the fix will work, waiting for CI to run).

mwcraig commented:
Problem fixed, astroquery 0.3.7 should be ready Thursday morning (central time).

mwcraig commented:
Just confirmed that astroquery 0.3.7 is available on the astropy conda channel.

eblur commented on tutorials/notebooks/FITS-cubes/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+astropy
+numpy
+matplotlib
+astroquery
```
Make a note that it requires astroquery 0.3.7 or above

adrn commented:
Hm, it's a bit awkward to leave comments on notebooks! So I'll just put comments here. I did ~quick read-throughs of each tutorial (but did not look for grammar / spelling mistakes). 

__General comments / questions__

* With the coordinates tutorials here, is the intention to replace the one that already exists with a series of shorter tutorials? If so, there is quite a bit of overlap with the existing tutorial, so we should probably make an issue to remove the duplicated material.

__coordinates-intro.ipynb__
* In first code cell under Exercise 1, remove the code - that seems to be the solution?

__coordinates-transform.ipynb__
* In the list of frames, two have extra quotation marks, e.g.: `GeocentricTrueEcliptic'`
* The comment `# c_icrs.fk5 does the same thing` doesn't make sense - there is no defined variable `c_icrs`! I would just remove it.
* "Transform frames to get to AltAz" - "AltAz" is never defined. Somewhere, it should say that this is shorthand for "altitude-azimuth"
* There is a not in a markdown cell `Today's date seems to be outside the limits of the IERS data, is this a problem??` - remove this!

__FITS-cubes.ipynb__
* This tutorial contains installation instructions, but none of the others do...Do we need to add similar instructions to all others, or remove this, or..?
* There seem to be some line-wrapping issues in the code cell under "Download the HI Fits Cube"
* Data file should be added to data.astropy.org so the link doesn't go dead!

__synthetic-images.ipynb__
* Data file should be added to data.astropy.org so the link doesn't go dead!

__units-and-integration.ipynb__
* String formatting of labels uses Python 2 style string formatting, e.g. `plt.xlabel(r'$\lambda \, [{\rm %s}]$' % lam.unit)`. This and others should be changed to, e.g.:
    ```
    plt.xlabel(r'$\lambda$ [{0:latex_inline}]'.format(lam.unit))
    ```
* The "Challenge problems" should be turned into distinct cells?

I have a bunch of minor suggestions that I can tweak in a follow-up PR.

eblur commented:
@eteq please add the data files from the **FITS-cubes** and **synthetic-images** tutorials to data.astropy.org

eblur commented:
> **General comments / questions**
> 
> With the coordinates tutorials here, is the intention to replace the one that already exists with a series of shorter tutorials? If so, there is quite a bit of overlap with the existing tutorial, so we should probably make an issue to remove the duplicated material.

Yes, we were breaking it apart into multiple tutorials. Mark as resolved.

> **coordinates-intro.ipynb**
> 
> In first code cell under Exercise 1, remove the code - that seems to be the solution?

Removed the answer. Mark as resolved.

> **coordinates-transform.ipynb**
>
> In the list of frames, two have extra quotation marks, e.g.: GeocentricTrueEcliptic'

Fixed.

> The comment # c_icrs.fk5 does the same thing doesn't make sense - there is no defined variable c_icrs! I would just remove it.

Fixed.

> "Transform frames to get to AltAz" - "AltAz" is never defined. Somewhere, it should say that this is shorthand for "altitude-azimuth"

Added words "altitude-azimuth" in the header and text description.

> There is a not in a markdown cell Today's date seems to be outside the limits of the IERS data, is this a problem?? - remove this!

Removed! Thanks for the catch.

> **FITS-cubes.ipynb**
> 
> This tutorial contains installation instructions, but none of the others do...Do we need to add similar instructions to all others, or remove this, or..?

The tutorial uses many packages that we do not typically ask people to install before workshops. I changed the instructions to use `conda` instead of `pip` because they are all available with conda.

> There seem to be some line-wrapping issues in the code cell under "Download the HI Fits Cube"

Don't see this problem when I open the notebook on my computer. I added some extra line breaks to give this code block some extra room.

> Data file should be added to data.astropy.org so the link doesn't go dead!

Need a reminder / help from @eteq 

> **synthetic-images.ipynb**
> 
> Data file should be added to data.astropy.org so the link doesn't go dead!

Need a reminder / help from @eteq 

> **units-and-integration.ipynb**
> 
> String formatting of labels uses Python 2 style string formatting, e.g. plt.xlabel(r'$\lambda \, [{\rm %s}]$' % lam.unit). This and others should be changed to, e.g.:
> plt.xlabel(r'$\lambda$ [{0:latex_inline}]'.format(lam.unit))

Done!

> The "Challenge problems" should be turned into distinct cells?

Skipping this for now.

eblur commented:
Hi @adrn please see my responses and mark the review as complete.

adrn commented:
It looks like this needs to be rebased on master - do you want to give that a try? I can help you through it, or I can just do it and push to your branch - your call!

One other question I have: if the coordinates tutorial has been split up, shouldn't we remove the original, long one?

adrn commented:
BTW: the travis build failed but didn't seem related to this PR, so I restarted it and it looks fine.

eblur commented:
All the tutorials run on my machine. I updated the requirements.txt files to reflect what versions they are running.

@eteq checked out some of the tutorials and found some broken code lines. We fixed one that was trivial.

I think all the Travis checks pass because the yml file chooses specific versions of packages that are slightly out of date now. Perhaps your (and @eteq's) installation is more up to date?

This highlights the need to fix the Travis infrastructure, see Issue #152 

If there are no major issues brought up in the next three days, I'll merge this PR.

adrn commented:
@eblur see my comments above! I think this PR needs to be rebased before it's merged...
